File: RepoTransBench/source_projects/Python/realpython_codetiming/tests/test_timer_additional.py
import math
import pytest
from codetiming._timer import Timer, TimerError

def test_timer_context_manager_runs():
    timer = Timer("cmmsg", text="Elapsed time: {0:.4f}s")
    with timer:
        pass  # just check context manager works

def test_timer_start_stop_elapsed():
    timer = Timer("simple", text="Time {0:.4f}")
    assert timer._start_time is None
    timer.start()
    assert timer._start_time is not None
    timer.stop()
    elapsed = timer.last
    assert isinstance(elapsed, float)

def test_timer_str_repr():
    timer = Timer("simple", text="Time {0:.4f}")
    assert isinstance(str(timer), str)
    assert "Timer" in repr(timer)

def test_timer_running_status_via_private():
    timer = Timer("runstat", text="Running:{0}")
    timer.start()
    assert timer._start_time is not None
    timer.stop()
    assert timer._start_time is None

def test_timer_logger_callable_text():
    messages = []
    timer = Timer("cbmsg", text=lambda s: f"Time={s:.2f}", logger=messages.append)
    timer.start()
    timer.stop()
    assert messages
    assert "Time=" in messages[0]

def test_timer_without_text_logger():
    timer = Timer("notext", text=None, logger=None)
    timer.start()
    timer.stop()  # Should simply not log or raise

def test_timer_stop_without_start_raises():
    timer = Timer("exception", text="fail")
    # Calling stop before start should raise TimerError
    with pytest.raises(TimerError):
        timer.stop()

def test_timer_multiple_starts_raises():
    timer = Timer("multi", text="multi")
    timer.start()
    with pytest.raises(TimerError):
        timer.start()
    timer.stop()

def test_timer_last_when_nan():
    timer = Timer("none", text="none")
    # Should return nan if never stopped
    assert math.isnan(timer.last)

def test_timer_compare_multiple_instances():
    timer1 = Timer("a")
    timer2 = Timer("b")
    assert timer1 is not timer2

File: RepoTransBench/target_projects/Python/Rust/realpython_codetiming/tests/original/test_timer_additional.rs
use realpython_codetiming::timer::{Timer, TimerText, TimerError};
// Panic/assertions in Rust will be used.
use std::rc::Rc;

#[test]
fn test_timer_context_manager_runs() {
    let mut timer = Timer::new("cmmsg", Some(TimerText::Plain("Elapsed time: {0:.4f}s".to_string())), None);
    // Context in Rust: start/stop explicitly
    timer.start().unwrap();
    timer.stop().unwrap();
}

#[test]
fn test_timer_start_stop_elapsed() {
    let mut timer = Timer::new("simple", Some(TimerText::Plain("Time {0:.4f}".to_string())), None);
    assert!(timer._start_time.is_none());
    timer.start().unwrap();
    assert!(timer._start_time.is_some());
    timer.stop().unwrap();
    let elapsed = timer.last;
    assert!(elapsed.is_finite());
}

#[test]
fn test_timer_str_repr() {
    let timer = Timer::new("simple", Some(TimerText::Plain("Time {0:.4f}".to_string())), None);
    let s = format!("{}", timer);
    let rep = format!("{:?}", timer);
    assert!(s.contains("Timer"));
    assert!(rep.contains("Timer"));
}

#[test]
fn test_timer_running_status_via_private() {
    let mut timer = Timer::new("runstat", Some(TimerText::Plain("Running:{0}".to_string())), None);
    timer.start().unwrap();
    assert!(timer._start_time.is_some());
    timer.stop().unwrap();
    assert!(timer._start_time.is_none());
}

#[test]
fn test_timer_logger_callable_text() {
    let messages = Rc::new(std::cell::RefCell::new(vec![]));
    let msg2 = Rc::clone(&messages);
    let logger = Rc::new(move |s: String| {
        msg2.borrow_mut().push(s);
    });
    let cb = Rc::new(|s: f64| format!("Time={:.2}", s));
    let mut timer = Timer::new("cbmsg", Some(TimerText::Function(cb)), Some(logger));
    timer.start().unwrap();
    timer.stop().unwrap();
    let msgs = messages.borrow();
    assert!(!msgs.is_empty());
    assert!(msgs[0].contains("Time="));
}

#[test]
fn test_timer_without_text_logger() {
    let mut timer =
        Timer::new("notext", Some(TimerText::None), None);
    timer.start().unwrap();
    timer.stop().unwrap();
}

#[test]
#[should_panic(expected = "TimerError: Timer is not running")]
fn test_timer_stop_without_start_raises() {
    let mut timer = Timer::new("exception", Some(TimerText::Plain("fail".to_string())), None);
    let _ = timer.stop().unwrap();
}

#[test]
fn test_timer_multiple_starts_raises() {
    let mut timer = Timer::new("multi", Some(TimerText::Plain("multi".to_string())), None);
    timer.start().unwrap();
    let double_start_result = timer.start();
    assert!(double_start_result.is_err());
    timer.stop().unwrap();
}

#[test]
fn test_timer_last_when_nan() {
    let timer = Timer::new("none", Some(TimerText::Plain("none".to_string())), None);
    assert!(timer.last.is_nan());
}

#[test]
fn test_timer_compare_multiple_instances() {
    let timer1 = Timer::new("a", None, None);
    let timer2 = Timer::new("b", None, None);
    // They should not compare as the same.
    assert!(timer1 != timer2);
}